File: src/main/scala/jk_5/nailed/server/scheduler/NailedScheduler.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.scheduler

import java.util
import java.util.concurrent
import java.util.concurrent.{Callable, TimeUnit}

import io.netty.util.concurrent._
import jk_5.eventbus.EventHandler
import jk_5.nailed.api.event.server.ServerPostTickEvent
import jk_5.nailed.api.scheduler.Scheduler

import scala.collection.mutable


/**
 * No description given
 *
 * @author jk-5
 */
object NailedScheduler extends Scheduler {
  val executor = new DefaultEventExecutorGroup(Runtime.getRuntime.availableProcessors() * 2)
  private val executionQueue = mutable.Queue[Runnable]()

  override def submit(task: Runnable): Future[_] = executor.submit(task)
  override def submit[T](task: Runnable, result: T): Future[T] = executor.submit(task, result)
  override def submit[T](task: Callable[T]): Future[T] = executor.submit(task)
  override def schedule(command: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture[_] = executor.schedule(command, delay, unit)
  override def schedule[V](callable: Callable[V], delay: Long, unit: TimeUnit): ScheduledFuture[V] = executor.schedule(callable, delay, unit)
  override def scheduleAtFixedRate(command: Runnable, initialDelay: Long, period: Long, unit: TimeUnit): ScheduledFuture[_] = executor.scheduleAtFixedRate(command, initialDelay, period, unit)
  override def scheduleWithFixedDelay(command: Runnable, initialDelay: Long, delay: Long, unit: TimeUnit): ScheduledFuture[_] = executor.scheduleWithFixedDelay(command, initialDelay, delay, unit)
  override def invokeAll[T](tasks: util.Collection[_ <: Callable[T]]): util.List[concurrent.Future[T]] = executor.invokeAll(tasks)
  override def invokeAll[T](tasks: util.Collection[_ <: Callable[T]], timeout: Long, unit: TimeUnit): util.List[concurrent.Future[T]] = executor.invokeAll(tasks, timeout, unit)
  override def invokeAny[T](tasks: util.Collection[_ <: Callable[T]]): T = executor.invokeAny(tasks)
  override def invokeAny[T](tasks: util.Collection[_ <: Callable[T]], timeout: Long, unit: TimeUnit): T = executor.invokeAny(tasks, timeout, unit)
  override def execute(command: Runnable): Unit = executor.execute(command)

  override def submitSync(task: Runnable) = this.submit(task, null)
  override def submitSync[T](task: Runnable, result: T): Future[T] = {
    val future = new DefaultPromise[T](this.executor.next())
    this.executionQueue += new Runnable {
      override def run(){
        try{
          task.run()
          future.setSuccess(result)
        }catch{
          case e: Exception => future.setFailure(e)
        }
      }
    }
    future
  }
  override def submitSync[T](task: Callable[T]): Future[T] = {
    val future = new DefaultPromise[T](this.executor.next())
    this.executionQueue += new Runnable {
      override def run(){
        try{
          future.setSuccess(task.call())
        }catch{
          case e: Exception => future.setFailure(e)
        }
      }
    }
    future
  }
  override def scheduleSync(command: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture[_] = ???
  override def scheduleSync[V](callable: Callable[V], delay: Long, unit: TimeUnit): ScheduledFuture[V] = ???
  override def scheduleAtFixedRateSync(command: Runnable, initialDelay: Long, period: Long, unit: TimeUnit): ScheduledFuture[_] = ???
  override def scheduleWithFixedDelaySync(command: Runnable, initialDelay: Long, delay: Long, unit: TimeUnit): ScheduledFuture[_] = ???
  override def invokeAllSync[T](tasks: util.Collection[_ <: Callable[T]]): util.List[concurrent.Future[T]] = ???
  override def invokeAllSync[T](tasks: util.Collection[_ <: Callable[T]], timeout: Long, unit: TimeUnit): util.List[concurrent.Future[T]] = ???
  override def invokeAnySync[T](tasks: util.Collection[_ <: Callable[T]]): T = ???
  override def invokeAnySync[T](tasks: util.Collection[_ <: Callable[T]], timeout: Long, unit: TimeUnit): T = ???
  override def executeSync(command: Runnable) = this.executionQueue += command

  @EventHandler
  def onTick(event: ServerPostTickEvent){
    if(executionQueue.size > 0){
      for(e <- executionQueue){
        e.run()
      }
      executionQueue.clear()
    }
  }

  override def shutdown(){}
  override def isTerminated = false
  override def awaitTermination(timeout: Long, unit: TimeUnit){}
  override def shutdownNow(){}
  override def isShutdown = false
}

File: src/main/java/jk_5/nailed/server/scheduler/NailedScheduler.java
package jk_5.nailed.server.scheduler;

import io.netty.util.concurrent.DefaultEventExecutorGroup;
import io.netty.util.concurrent.DefaultPromise;
import jk_5.eventbus.EventHandler;
import jk_5.nailed.api.event.server.ServerPostTickEvent;
import jk_5.nailed.api.scheduler.Scheduler;

import javax.annotation.Nonnull;
import java.util.*;
import java.util.concurrent.*;

public class NailedScheduler implements Scheduler {

    private static final NailedScheduler INSTANCE = new NailedScheduler();

    private final DefaultEventExecutorGroup executor = new DefaultEventExecutorGroup(Runtime.getRuntime().availableProcessors() * 2);
    private final Queue<Runnable> executionQueue = new LinkedList<Runnable>();

    @Nonnull
    @Override
    public Future<?> submit(@Nonnull Runnable task) {
        return executor.submit(task);
    }

    @Nonnull
    @Override
    public <T> Future<T> submit(@Nonnull Runnable task, T result) {
        return executor.submit(task, result);
    }

    @Nonnull
    @Override
    public <T> Future<T> submit(@Nonnull Callable<T> task) {
        return executor.submit(task);
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> schedule(@Nonnull Runnable command, long delay, @Nonnull TimeUnit unit) {
        return executor.schedule(command, delay, unit);
    }

    @Nonnull
    @Override
    public <V> ScheduledFuture<V> schedule(@Nonnull Callable<V> callable, long delay, @Nonnull TimeUnit unit) {
        return executor.schedule(callable, delay, unit);
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> scheduleAtFixedRate(@Nonnull Runnable command, long initialDelay, long period, @Nonnull TimeUnit unit) {
        return executor.scheduleAtFixedRate(command, initialDelay, period, unit);
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> scheduleWithFixedDelay(@Nonnull Runnable command, long initialDelay, long delay, @Nonnull TimeUnit unit) {
        return executor.scheduleWithFixedDelay(command, initialDelay, delay, unit);
    }

    @Nonnull
    @Override
    public <T> List<Future<T>> invokeAll(@Nonnull Collection<? extends Callable<T>> tasks) throws InterruptedException {
        return executor.invokeAll(tasks);
    }

    @Nonnull
    @Override
    public <T> List<Future<T>> invokeAll(@Nonnull Collection<? extends Callable<T>> tasks, long timeout, @Nonnull TimeUnit unit) throws InterruptedException {
        return executor.invokeAll(tasks, timeout, unit);
    }

    @Nonnull
    @Override
    public <T> T invokeAny(@Nonnull Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
        return executor.invokeAny(tasks);
    }

    @Override
    public <T> T invokeAny(@Nonnull Collection<? extends Callable<T>> tasks, long timeout, @Nonnull TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        return executor.invokeAny(tasks, timeout, unit);
    }

    @Override
    public void execute(@Nonnull Runnable command) {
        executor.execute(command);
    }

    @Nonnull
    @Override
    public Future<?> submitSync(@Nonnull Runnable task) {
        return this.submitSync(task, null);
    }

    @Nonnull
    @Override
    public <T> Future<T> submitSync(@Nonnull final Runnable task, final T result) {
        final DefaultPromise<T> future = new DefaultPromise<T>(this.executor.next());
        this.executionQueue.add(new Runnable() {
            @Override
            public void run() {
                try{
                    task.run();
                    future.setSuccess(result);
                }catch(Exception e){
                    future.setFailure(e);
                }
            }
        });
        return future;
    }

    @Nonnull
    @Override
    public <T> Future<T> submitSync(@Nonnull final Callable<T> task) {
        final DefaultPromise<T> future = new DefaultPromise<T>(this.executor.next());
        this.executionQueue.add(new Runnable() {
            @Override
            public void run() {
                try{
                    future.setSuccess(task.call());
                }catch(Exception e){
                    future.setFailure(e);
                }
            }
        });
        return future;
    }

    @Override
    public <T> T invokeAnySync(@Nonnull Collection<? extends Callable<T>> tasks, long timeout, @Nonnull TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> scheduleSync(@Nonnull Runnable command, long delay, @Nonnull TimeUnit unit) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public <V> ScheduledFuture<V> scheduleSync(@Nonnull Callable<V> callable, long delay, @Nonnull TimeUnit unit) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> scheduleAtFixedRateSync(@Nonnull Runnable command, long initialDelay, long period, @Nonnull TimeUnit unit) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public ScheduledFuture<?> scheduleWithFixedDelaySync(@Nonnull Runnable command, long initialDelay, long delay, @Nonnull TimeUnit unit) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public <T> List<Future<T>> invokeAllSync(@Nonnull Collection<? extends Callable<T>> tasks) throws InterruptedException {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public <T> List<Future<T>> invokeAllSync(@Nonnull Collection<? extends Callable<T>> tasks, long timeout, @Nonnull TimeUnit unit) throws InterruptedException {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public <T> T invokeAnySync(@Nonnull Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
        throw new UnsupportedOperationException();
    }

    @Override
    public void executeSync(@Nonnull Runnable command) {
        this.executionQueue.add(command);
    }

    @Override
    public void shutdown() {

    }

    @Override
    public List<Runnable> shutdownNow() {
        return Collections.emptyList();
    }

    @Override
    public boolean isShutdown() {
        return false;
    }

    @Override
    public boolean isTerminated() {
        return false;
    }

    @Override
    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
        return false;
    }

    @EventHandler
    public void onTick(ServerPostTickEvent event){
        if(executionQueue.size() > 0){
            for (Runnable e : executionQueue) {
                e.run();
            }
            executionQueue.clear();
        }
    }

    public DefaultEventExecutorGroup getExecutor() {
        return executor;
    }

    public static NailedScheduler instance(){
        return INSTANCE;
    }
}
